{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0fe15c86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/methylation/F24A080000424_MUSekgzH_20240805100100'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "69aab01c-e8ff-4ae4-be24-4f22441b8b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "root_dir=\"/methylation/F24A080000424_MUSekgzH_20240805100100/\"  # 基因文件的根目录\n",
    "sample_names=[\"13A\",\"32A\",\"50A\",\"26A\",\"28A\",\"38A\",\"74A\",\"25A\",\"75A\",\"88A\",\"97A\"]     # 样本名称\n",
    "\n",
    "# 报告存放的文件夹\n",
    "if not os.path.exists(f\"{root_dir}report/\"):\n",
    "    os.makedirs(f\"{root_dir}report/\")\n",
    "for sample_name in sample_names:\n",
    "    if not os.path.exists(f\"{root_dir}report/{sample_name}\"):\n",
    "        os.makedirs(f\"{root_dir}report/{sample_name}\")\n",
    "# 程序数据输出的文件夹\n",
    "if not os.path.exists(f\"{root_dir}output/\"):\n",
    "    os.makedirs(f\"{root_dir}output/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a882cc0b-f70e-4c77-9437-deae47924d85",
   "metadata": {},
   "source": [
    "## 1. 数据基本处理与质控\n",
    "将下机数据进行过滤，包括去污染，去测序接头和低质量碱基比例过高的reads，得到clean data。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1a09b238",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clean Reads</th>\n",
       "      <th>Clean Base</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>Q20(%)</th>\n",
       "      <th>Q30(%)</th>\n",
       "      <th>GC(%)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13A</th>\n",
       "      <td>360085078</td>\n",
       "      <td>108025523400</td>\n",
       "      <td>PE150</td>\n",
       "      <td>95.58</td>\n",
       "      <td>90.74</td>\n",
       "      <td>22.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25A</th>\n",
       "      <td>360265525</td>\n",
       "      <td>108079657500</td>\n",
       "      <td>PE150</td>\n",
       "      <td>96.62</td>\n",
       "      <td>89.73</td>\n",
       "      <td>22.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26A</th>\n",
       "      <td>360191949</td>\n",
       "      <td>108057584700</td>\n",
       "      <td>PE150</td>\n",
       "      <td>96.84</td>\n",
       "      <td>90.23</td>\n",
       "      <td>22.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28A</th>\n",
       "      <td>360244786</td>\n",
       "      <td>108073435800</td>\n",
       "      <td>PE150</td>\n",
       "      <td>95.83</td>\n",
       "      <td>91.18</td>\n",
       "      <td>22.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32A</th>\n",
       "      <td>360173146</td>\n",
       "      <td>108051943800</td>\n",
       "      <td>PE150</td>\n",
       "      <td>96.72</td>\n",
       "      <td>92.74</td>\n",
       "      <td>22.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38A</th>\n",
       "      <td>360332442</td>\n",
       "      <td>108099732600</td>\n",
       "      <td>PE150</td>\n",
       "      <td>97.92</td>\n",
       "      <td>95.14</td>\n",
       "      <td>21.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50A</th>\n",
       "      <td>360257577</td>\n",
       "      <td>108077273100</td>\n",
       "      <td>PE150</td>\n",
       "      <td>97.89</td>\n",
       "      <td>95.11</td>\n",
       "      <td>21.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74A</th>\n",
       "      <td>360194122</td>\n",
       "      <td>108058236600</td>\n",
       "      <td>PE150</td>\n",
       "      <td>95.39</td>\n",
       "      <td>88.52</td>\n",
       "      <td>21.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75A</th>\n",
       "      <td>323495667</td>\n",
       "      <td>97048700100</td>\n",
       "      <td>PE150</td>\n",
       "      <td>95.07</td>\n",
       "      <td>87.86</td>\n",
       "      <td>22.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88A</th>\n",
       "      <td>354517702</td>\n",
       "      <td>106355310600</td>\n",
       "      <td>PE150</td>\n",
       "      <td>94.92</td>\n",
       "      <td>87.76</td>\n",
       "      <td>22.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97A</th>\n",
       "      <td>360119067</td>\n",
       "      <td>108035720100</td>\n",
       "      <td>PE150</td>\n",
       "      <td>95.46</td>\n",
       "      <td>88.78</td>\n",
       "      <td>22.82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Clean Reads    Clean Base Read Length  Q20(%)  Q30(%)  GC(%)\n",
       "Sample Name                                                              \n",
       "13A            360085078  108025523400       PE150   95.58   90.74  22.42\n",
       "25A            360265525  108079657500       PE150   96.62   89.73  22.86\n",
       "26A            360191949  108057584700       PE150   96.84   90.23  22.36\n",
       "28A            360244786  108073435800       PE150   95.83   91.18  22.61\n",
       "32A            360173146  108051943800       PE150   96.72   92.74  22.68\n",
       "38A            360332442  108099732600       PE150   97.92   95.14  21.63\n",
       "50A            360257577  108077273100       PE150   97.89   95.11  21.61\n",
       "74A            360194122  108058236600       PE150   95.39   88.52  21.66\n",
       "75A            323495667   97048700100       PE150   95.07   87.86  22.01\n",
       "88A            354517702  106355310600       PE150   94.92   87.76  22.75\n",
       "97A            360119067  108035720100       PE150   95.46   88.78  22.82"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 表1 测序质量基本统计（公司提供）\n",
    "report_table1=pd.read_csv(f'{root_dir}Clean_stat.xls', sep='\\t', index_col=0)\n",
    "report_table1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c48056b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 图1：测序碱基含量分布\n",
    "# 图2：碱基测序质量分布情况\n",
    "# 公司已提供，位于每个样本文件夹中。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6ab430b6-f1f9-48de-98d6-0fe48231c86f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Id</th>\n",
       "      <th>Clean Reads</th>\n",
       "      <th>Uniquely Mapped Reads</th>\n",
       "      <th>Uniquely Mapping Rate</th>\n",
       "      <th>Bisulfite Conversion Rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13A</td>\n",
       "      <td>360085078</td>\n",
       "      <td>258302605</td>\n",
       "      <td>71.73</td>\n",
       "      <td>99.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32A</td>\n",
       "      <td>360173146</td>\n",
       "      <td>279886483</td>\n",
       "      <td>77.71</td>\n",
       "      <td>99.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50A</td>\n",
       "      <td>360257577</td>\n",
       "      <td>302777677</td>\n",
       "      <td>84.04</td>\n",
       "      <td>99.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26A</td>\n",
       "      <td>360191949</td>\n",
       "      <td>272931733</td>\n",
       "      <td>75.77</td>\n",
       "      <td>99.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28A</td>\n",
       "      <td>360244786</td>\n",
       "      <td>257555160</td>\n",
       "      <td>71.49</td>\n",
       "      <td>99.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>38A</td>\n",
       "      <td>360332442</td>\n",
       "      <td>297628490</td>\n",
       "      <td>82.60</td>\n",
       "      <td>99.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>74A</td>\n",
       "      <td>360194122</td>\n",
       "      <td>280008801</td>\n",
       "      <td>77.74</td>\n",
       "      <td>99.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>25A</td>\n",
       "      <td>360265525</td>\n",
       "      <td>278478393</td>\n",
       "      <td>77.30</td>\n",
       "      <td>99.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>75A</td>\n",
       "      <td>323495667</td>\n",
       "      <td>218869806</td>\n",
       "      <td>67.66</td>\n",
       "      <td>99.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>88A</td>\n",
       "      <td>354517702</td>\n",
       "      <td>261768036</td>\n",
       "      <td>73.84</td>\n",
       "      <td>99.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>97A</td>\n",
       "      <td>360119067</td>\n",
       "      <td>259451502</td>\n",
       "      <td>72.05</td>\n",
       "      <td>99.38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample Id  Clean Reads  Uniquely Mapped Reads  Uniquely Mapping Rate  \\\n",
       "0        13A    360085078              258302605                  71.73   \n",
       "1        32A    360173146              279886483                  77.71   \n",
       "2        50A    360257577              302777677                  84.04   \n",
       "3        26A    360191949              272931733                  75.77   \n",
       "4        28A    360244786              257555160                  71.49   \n",
       "5        38A    360332442              297628490                  82.60   \n",
       "6        74A    360194122              280008801                  77.74   \n",
       "7        25A    360265525              278478393                  77.30   \n",
       "8        75A    323495667              218869806                  67.66   \n",
       "9        88A    354517702              261768036                  73.84   \n",
       "10       97A    360119067              259451502                  72.05   \n",
       "\n",
       "    Bisulfite Conversion Rate  \n",
       "0                       99.44  \n",
       "1                       99.43  \n",
       "2                       99.47  \n",
       "3                       99.44  \n",
       "4                       99.44  \n",
       "5                       99.46  \n",
       "6                       99.44  \n",
       "7                       99.42  \n",
       "8                       99.39  \n",
       "9                       99.40  \n",
       "10                      99.38  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 表2 比对结果统计\n",
    "\n",
    "report_table2=[]\n",
    "for sample_name in sample_names:\n",
    "    data={}\n",
    "    with open(f\"{root_dir}{sample_name}/output/bismark_alignment/{sample_name}_1_bismark_bt2_PE_report.txt\", \"r\") as file:\n",
    "        for line in file:\n",
    "            if line.count(\":\") == 1:   # 匹配包含1个冒号的行\n",
    "                parts = line.split(\":\")\n",
    "                try:\n",
    "                    data[parts[0].strip()]=int(parts[1].strip())\n",
    "                except:\n",
    "                    pass\n",
    "    item={}\n",
    "    item[\"Sample Id\"]=sample_name\n",
    "    item[\"Clean Reads\"]=data[\"Sequence pairs analysed in total\"]\n",
    "    item[\"Uniquely Mapped Reads\"]=data[\"Number of paired-end alignments with a unique best hit\"]\n",
    "    item[\"Uniquely Mapping Rate\"]=round(data[\"Number of paired-end alignments with a unique best hit\"]/data[\"Sequence pairs analysed in total\"]*100,2)\n",
    "\n",
    "    # Bisulfite Conversion Rate的计算规则存疑\n",
    "    # item[\"Bisulfite Conversion Rate\"]=round((data[\"Total unmethylated C's in CpG context\"]+data[\"Total unmethylated C's in CHG context\"]+data[\"Total unmethylated C's in CHH context\"])/data[\"Total number of C's analysed\"]*100,2)\n",
    "    item[\"Bisulfite Conversion Rate\"]=round((data[\"Total unmethylated C's in CHG context\"]+data[\"Total unmethylated C's in CHH context\"])/(data[\"Total unmethylated C's in CHG context\"]+data[\"Total unmethylated C's in CHH context\"]+data[\"Total methylated C's in CHG context\"]+data[\"Total methylated C's in CHH context\"]+data[\"Total methylated C's in Unknown context\"])*100,2)\n",
    "\n",
    "    report_table2.append(item)\n",
    "report_table2=pd.DataFrame(report_table2)\n",
    "report_table2.to_csv(f\"{root_dir}report/QC quality control table for each sample.tsv\",sep=\"\\t\",index=False)\n",
    "report_table2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ad95fdc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 图3 测序深度分布图\n",
    "\n",
    "def plot_methylation_depth_distribution(sample_name):\n",
    "    input_file=f\"{root_dir}{sample_name}/output/{sample_name}_methylation_depth_report.txt\"\n",
    "    output_file=f\"{root_dir}report/{sample_name}/Coverage of Corresponding Depth in Sample.jpg\"\n",
    "    df = pd.read_csv(input_file,sep=\"\\t\")\n",
    "    df[\"C\"]=df.sum(axis=1)\n",
    "    df=df.set_index(\"Depth\")\n",
    "    df=df[[\"C\",\"CG\",\"CHG\",\"CHH\"]]\n",
    "    df=df/df.sum(axis=0)\n",
    "    df.plot()\n",
    "    plt.title(f\"Coverage of Corresponding Depth in Sample:{sample_name}\")\n",
    "    plt.ylabel(\"Fraction of Covered (%)\")\n",
    "    plt.legend(title='Context')\n",
    "    plt.savefig(output_file,dpi=1000)\n",
    "    plt.close()\n",
    "\n",
    "for sample_name in sample_names:\n",
    "    df_coverage=plot_methylation_depth_distribution(sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "637d7436",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 图4 C碱基测序深度的累积分布图\n",
    "def plot_cumulative_methylation_depth_distribution(sample_name):\n",
    "    input_file=f\"{root_dir}{sample_name}/output/{sample_name}_methylation_depth_report.txt\"\n",
    "    output_file=f\"{root_dir}report/{sample_name}/Cumulative Coverage of Corresponding Depth in Sample.jpg\"\n",
    "    df = pd.read_csv(input_file,sep=\"\\t\")\n",
    "    df[\"C\"]=df.sum(axis=1)\n",
    "    df=df.set_index(\"Depth\")\n",
    "    df=df[[\"C\",\"CG\",\"CHG\",\"CHH\"]]\n",
    "    df=df/df.sum(axis=0)\n",
    "    df = df.cumsum() # 计算累积分布\n",
    "    df.plot()\n",
    "    plt.title(f\"Cumulative Coverage of Corresponding Depth in Sample:{sample_name}\")\n",
    "    plt.ylabel(\"Fraction of Covered (%)\")\n",
    "    plt.legend(title='Context')\n",
    "    plt.savefig(output_file,dpi=1000)\n",
    "    plt.close()\n",
    "    \n",
    "for sample_name in sample_names:\n",
    "    df_coverage=plot_cumulative_methylation_depth_distribution(sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5dc979d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>C (%)</th>\n",
       "      <th>CG (%)</th>\n",
       "      <th>CHG (%)</th>\n",
       "      <th>CHH (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>96.04</td>\n",
       "      <td>96.50</td>\n",
       "      <td>97.51</td>\n",
       "      <td>95.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000068.8</td>\n",
       "      <td>95.44</td>\n",
       "      <td>96.28</td>\n",
       "      <td>96.91</td>\n",
       "      <td>94.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000069.7</td>\n",
       "      <td>95.80</td>\n",
       "      <td>96.20</td>\n",
       "      <td>97.22</td>\n",
       "      <td>95.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000070.7</td>\n",
       "      <td>94.51</td>\n",
       "      <td>95.22</td>\n",
       "      <td>95.94</td>\n",
       "      <td>94.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000071.7</td>\n",
       "      <td>94.90</td>\n",
       "      <td>96.13</td>\n",
       "      <td>96.52</td>\n",
       "      <td>94.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000084.7</td>\n",
       "      <td>96.35</td>\n",
       "      <td>97.05</td>\n",
       "      <td>97.84</td>\n",
       "      <td>95.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000085.7</td>\n",
       "      <td>96.86</td>\n",
       "      <td>97.80</td>\n",
       "      <td>98.25</td>\n",
       "      <td>96.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000086.8</td>\n",
       "      <td>88.95</td>\n",
       "      <td>87.82</td>\n",
       "      <td>89.95</td>\n",
       "      <td>88.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>0.78</td>\n",
       "      <td>1.35</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>90.30</td>\n",
       "      <td>90.59</td>\n",
       "      <td>91.93</td>\n",
       "      <td>90.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>242 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name          Chr  C (%)  CG (%)  CHG (%)  CHH (%)\n",
       "0           13A  NC_000067.7  96.04   96.50    97.51    95.61\n",
       "1           13A  NC_000068.8  95.44   96.28    96.91    94.98\n",
       "2           13A  NC_000069.7  95.80   96.20    97.22    95.39\n",
       "3           13A  NC_000070.7  94.51   95.22    95.94    94.06\n",
       "4           13A  NC_000071.7  94.90   96.13    96.52    94.36\n",
       "..          ...          ...    ...     ...      ...      ...\n",
       "237         97A  NC_000084.7  96.35   97.05    97.84    95.90\n",
       "238         97A  NC_000085.7  96.86   97.80    98.25    96.41\n",
       "239         97A  NC_000086.8  88.95   87.82    89.95    88.74\n",
       "240         97A  NC_000087.8   0.78    1.35     0.80     0.75\n",
       "241         97A  NC_005089.1  90.30   90.59    91.93    90.00\n",
       "\n",
       "[242 rows x 6 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 表3、表4 样品在全基因组各染色体上的C位点覆盖度\n",
    "def calc_coverage_rate_by_chromosome(sample_name):\n",
    "    input_file=f\"{root_dir}{sample_name}/output/{sample_name}_methylation_coverage_report.txt\"\n",
    "    output_file=f\"{root_dir}report/{sample_name}/Coverage Rate Group By Chromosome.tsv\"\n",
    "    df = pd.read_csv(input_file,sep=\"\\t\")\n",
    "    # 过滤只保留Chromosome列中以NC开头的行\n",
    "    df = df[df['Chromosome'].str.startswith('NC')]\n",
    "    # 计算每个 context 的覆盖率\n",
    "    df['CoverageRate'] = round(df['covered'] / df['Count']*100,2)\n",
    "\n",
    "    # 计算每个染色体的整体覆盖率\n",
    "    overall_coverage = df.groupby('Chromosome').apply(lambda x: round(x['covered'].sum() / x['Count'].sum()*100,2),include_groups=False).reset_index()\n",
    "    overall_coverage.columns = ['Chromosome', 'C (%)']\n",
    "\n",
    "    # 创建以染色体为index、context为表头的透视表，并添加总体覆盖率列\n",
    "    pivot_table = df.pivot(index='Chromosome', columns='Context', values='CoverageRate')\n",
    "\n",
    "    # 将整体覆盖率加入透视表\n",
    "    pivot_table = overall_coverage.merge(pivot_table, on='Chromosome')\n",
    "\n",
    "    pivot_table.columns=[\"Chr\",\"C (%)\", \"CG (%)\", \"CHG (%)\", \"CHH (%)\"]\n",
    "\n",
    "    # 保存结果到文件\n",
    "    pivot_table.to_csv(output_file,sep=\"\\t\",index=False)\n",
    "    return pivot_table\n",
    "df_coverage_list=[]\n",
    "for sample_name in sample_names:\n",
    "    df_coverage=calc_coverage_rate_by_chromosome(sample_name)\n",
    "    df_coverage.insert(0,\"sample_name\",sample_name)\n",
    "    df_coverage_list.append(df_coverage)\n",
    "df_coverage_list=pd.concat(df_coverage_list,axis=0).reset_index(drop=True)\n",
    "df_coverage_list.to_csv(f\"{root_dir}report/Coverage Rate Group By Chromosome.tsv\",sep=\"\\t\",index=False)\n",
    "df_coverage_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cc0ebe07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 表5、表6 样品在全基因组各类型调控元件范围内的C位点覆盖度\n",
    "# 暂时不做"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bd962273",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Clean Data Size (bp)</th>\n",
       "      <th>Mapping Rate (%)</th>\n",
       "      <th>Bisulfite Conversion Rate</th>\n",
       "      <th>Duplication Rate (%)</th>\n",
       "      <th>Average Depth (X)</th>\n",
       "      <th>Coverage (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13A</td>\n",
       "      <td>360085078</td>\n",
       "      <td>71.73</td>\n",
       "      <td>77.43</td>\n",
       "      <td>10.06</td>\n",
       "      <td>10.81</td>\n",
       "      <td>92.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32A</td>\n",
       "      <td>360173146</td>\n",
       "      <td>77.71</td>\n",
       "      <td>77.43</td>\n",
       "      <td>9.89</td>\n",
       "      <td>11.86</td>\n",
       "      <td>92.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50A</td>\n",
       "      <td>360257577</td>\n",
       "      <td>84.04</td>\n",
       "      <td>78.40</td>\n",
       "      <td>5.15</td>\n",
       "      <td>13.12</td>\n",
       "      <td>92.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26A</td>\n",
       "      <td>360191949</td>\n",
       "      <td>75.77</td>\n",
       "      <td>77.59</td>\n",
       "      <td>6.78</td>\n",
       "      <td>12.24</td>\n",
       "      <td>92.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28A</td>\n",
       "      <td>360244786</td>\n",
       "      <td>71.49</td>\n",
       "      <td>77.27</td>\n",
       "      <td>8.85</td>\n",
       "      <td>10.95</td>\n",
       "      <td>91.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>38A</td>\n",
       "      <td>360332442</td>\n",
       "      <td>82.60</td>\n",
       "      <td>78.39</td>\n",
       "      <td>5.54</td>\n",
       "      <td>12.91</td>\n",
       "      <td>93.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>74A</td>\n",
       "      <td>360194122</td>\n",
       "      <td>77.74</td>\n",
       "      <td>78.19</td>\n",
       "      <td>5.06</td>\n",
       "      <td>12.13</td>\n",
       "      <td>92.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>25A</td>\n",
       "      <td>360265525</td>\n",
       "      <td>77.30</td>\n",
       "      <td>77.01</td>\n",
       "      <td>5.90</td>\n",
       "      <td>13.06</td>\n",
       "      <td>92.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>75A</td>\n",
       "      <td>323495667</td>\n",
       "      <td>67.66</td>\n",
       "      <td>77.50</td>\n",
       "      <td>4.95</td>\n",
       "      <td>9.72</td>\n",
       "      <td>91.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>88A</td>\n",
       "      <td>354517702</td>\n",
       "      <td>73.84</td>\n",
       "      <td>76.91</td>\n",
       "      <td>4.75</td>\n",
       "      <td>12.00</td>\n",
       "      <td>92.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>97A</td>\n",
       "      <td>360119067</td>\n",
       "      <td>72.05</td>\n",
       "      <td>76.85</td>\n",
       "      <td>5.55</td>\n",
       "      <td>11.54</td>\n",
       "      <td>92.15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample ID  Clean Data Size (bp)  Mapping Rate (%)  \\\n",
       "0        13A             360085078             71.73   \n",
       "1        32A             360173146             77.71   \n",
       "2        50A             360257577             84.04   \n",
       "3        26A             360191949             75.77   \n",
       "4        28A             360244786             71.49   \n",
       "5        38A             360332442             82.60   \n",
       "6        74A             360194122             77.74   \n",
       "7        25A             360265525             77.30   \n",
       "8        75A             323495667             67.66   \n",
       "9        88A             354517702             73.84   \n",
       "10       97A             360119067             72.05   \n",
       "\n",
       "    Bisulfite Conversion Rate Duplication Rate (%)  Average Depth (X)  \\\n",
       "0                       77.43                10.06              10.81   \n",
       "1                       77.43                 9.89              11.86   \n",
       "2                       78.40                 5.15              13.12   \n",
       "3                       77.59                 6.78              12.24   \n",
       "4                       77.27                 8.85              10.95   \n",
       "5                       78.39                 5.54              12.91   \n",
       "6                       78.19                 5.06              12.13   \n",
       "7                       77.01                 5.90              13.06   \n",
       "8                       77.50                 4.95               9.72   \n",
       "9                       76.91                 4.75              12.00   \n",
       "10                      76.85                 5.55              11.54   \n",
       "\n",
       "    Coverage (%)  \n",
       "0          92.07  \n",
       "1          92.31  \n",
       "2          92.75  \n",
       "3          92.58  \n",
       "4          91.71  \n",
       "5          93.38  \n",
       "6          92.14  \n",
       "7          92.64  \n",
       "8          91.89  \n",
       "9          92.12  \n",
       "10         92.15  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 表7 各样品QC质控表(去重之后的数据)\n",
    "def get_report_table7(sample_name):\n",
    "    import re\n",
    "    PE_report={}\n",
    "    with open(f\"{root_dir}{sample_name}/output/bismark_alignment/{sample_name}_1_bismark_bt2_PE_report.txt\", \"r\") as file:\n",
    "        for line in file:\n",
    "            if line.count(\":\") == 1:   # 匹配包含1个冒号的行\n",
    "                parts = line.split(\":\")\n",
    "                try:\n",
    "                    PE_report[parts[0].strip()]=int(parts[1].strip())\n",
    "                except:\n",
    "                    pass\n",
    "    \n",
    "    # 查找deduplication_percentage\n",
    "    percentage_pattern = re.compile(r\"Total number duplicated alignments removed:.*?(\\d+\\.\\d+)%\")\n",
    "    with open(f\"{root_dir}{sample_name}/output/bismark_deduplicate/{sample_name}_1_bismark_bt2_pe.deduplication_report.txt\", \"r\") as file:\n",
    "        for line in file:\n",
    "            match = percentage_pattern.search(line)\n",
    "            if match:\n",
    "                deduplication_percentage = match.group(1)\n",
    "                break\n",
    "    \n",
    "    # 查找去重后的甲基化、非甲基化数据\n",
    "    splitting_report={}\n",
    "    with open(f\"{root_dir}{sample_name}/output/bismark_methylation/{sample_name}_1_bismark_bt2_pe.deduplicated_splitting_report.txt\", \"r\") as file:\n",
    "        for line in file:\n",
    "            if line.count(\":\") == 1:   # 匹配包含1个冒号的行\n",
    "                parts = line.split(\":\")\n",
    "                try:\n",
    "                    splitting_report[parts[0].strip()]=int(parts[1].strip())\n",
    "                except:\n",
    "                    pass\n",
    "    \n",
    "    # 计算平均深度\n",
    "    depth_report=pd.read_csv(f\"{root_dir}{sample_name}/output/{sample_name}_methylation_depth_report.txt\",sep=\"\\t\")\n",
    "    depth_report[\"C\"]=depth_report[[\"CG\",\"CHG\",\"CHH\"]].sum(axis=1)\n",
    "    depth_report[\"CxDepth\"]=depth_report[\"C\"]*depth_report[\"Depth\"]\n",
    "    avg_depth = (depth_report[\"CxDepth\"].sum()/depth_report[\"C\"].sum()).round(2)\n",
    "\n",
    "    # 计算Coverage\n",
    "    coverage_report=pd.read_csv(f\"{root_dir}{sample_name}/output/{sample_name}_methylation_coverage_report.txt\",sep=\"\\t\")\n",
    "    coverage=(coverage_report[\"covered\"].sum()/coverage_report[\"Count\"].sum()*100).round(2)\n",
    "\n",
    "    item={}\n",
    "    item[\"Sample ID\"]=sample_name\n",
    "    item[\"Clean Data Size (bp)\"]=PE_report[\"Sequence pairs analysed in total\"]\n",
    "    item[\"Mapping Rate (%)\"]=round(PE_report[\"Number of paired-end alignments with a unique best hit\"]/PE_report[\"Sequence pairs analysed in total\"]*100,2)\n",
    "\n",
    "    # Bisulfite Conversion Rate的计算规则存疑\n",
    "    # item[\"Bisulfite Conversion Rate (%)\"]=round((splitting_report[\"Total C to T conversions in CpG context\"]+splitting_report[\"Total C to T conversions in CHG context\"]+splitting_report[\"Total C to T conversions in CHH context\"])/splitting_report[\"Total number of C's analysed\"]*100,2)\n",
    "    item[\"Bisulfite Conversion Rate\"]=round((splitting_report[\"Total C to T conversions in CpG context\"]+splitting_report[\"Total C to T conversions in CHH context\"])/(splitting_report[\"Total C to T conversions in CHG context\"]+splitting_report[\"Total C to T conversions in CHH context\"]+splitting_report[\"Total methylated C's in CHG context\"]+splitting_report[\"Total methylated C's in CHH context\"])*100,2)\n",
    "    item[\"Duplication Rate (%)\"]=deduplication_percentage\n",
    "    item[\"Average Depth (X)\"]=avg_depth\n",
    "    item[\"Coverage (%)\"]=coverage\n",
    "    return item\n",
    "\n",
    "report_table7=[]\n",
    "for sample_name in sample_names:\n",
    "    item = get_report_table7(sample_name)\n",
    "    report_table7.append(item)\n",
    "report_table7=pd.DataFrame(report_table7)\n",
    "report_table7.to_csv(f\"{root_dir}report/QC quality control table for each sample (deduplicated).tsv\",sep=\"\\t\",index=False)\n",
    "report_table7\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6ef860d",
   "metadata": {},
   "source": [
    "##  2. 全基因组甲基化水平分析\n",
    "用于分析的DNA样品为多细胞样品，因此C碱基的甲基化水平是一个0% ～ 100%范围内的数值，等于该C碱基上覆盖到的支持mC的序列数除以有效覆盖的序列总数，通常CG甲基化存在于基因和重复序列中，在基因表达调控过程中起到非常重要的作用。非CG类型的序列（CHG和CHH）在基因中十分少见，主要存在于基因间区和富含重复序列的区域，在沉默转座子过程中起关键作用。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1615e875",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>C (%)</th>\n",
       "      <th>CG (%)</th>\n",
       "      <th>CHG (%)</th>\n",
       "      <th>CHH (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>3.50</td>\n",
       "      <td>75.93</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000068.8</td>\n",
       "      <td>3.84</td>\n",
       "      <td>75.19</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000069.7</td>\n",
       "      <td>3.53</td>\n",
       "      <td>75.84</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000070.7</td>\n",
       "      <td>3.74</td>\n",
       "      <td>74.62</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13A</td>\n",
       "      <td>NC_000071.7</td>\n",
       "      <td>3.92</td>\n",
       "      <td>75.08</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000084.7</td>\n",
       "      <td>4.09</td>\n",
       "      <td>79.04</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000085.7</td>\n",
       "      <td>4.35</td>\n",
       "      <td>77.00</td>\n",
       "      <td>0.63</td>\n",
       "      <td>0.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000086.8</td>\n",
       "      <td>3.12</td>\n",
       "      <td>77.14</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>9.46</td>\n",
       "      <td>83.24</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>97A</td>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.43</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.48</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>242 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name          Chr  C (%)  CG (%)  CHG (%)  CHH (%)\n",
       "0           13A  NC_000067.7   3.50   75.93     0.54     0.52\n",
       "1           13A  NC_000068.8   3.84   75.19     0.55     0.54\n",
       "2           13A  NC_000069.7   3.53   75.84     0.54     0.52\n",
       "3           13A  NC_000070.7   3.74   74.62     0.55     0.53\n",
       "4           13A  NC_000071.7   3.92   75.08     0.55     0.53\n",
       "..          ...          ...    ...     ...      ...      ...\n",
       "237         97A  NC_000084.7   4.09   79.04     0.59     0.59\n",
       "238         97A  NC_000085.7   4.35   77.00     0.63     0.64\n",
       "239         97A  NC_000086.8   3.12   77.14     0.54     0.52\n",
       "240         97A  NC_000087.8   9.46   83.24     0.64     0.57\n",
       "241         97A  NC_005089.1   0.47    0.43     0.47     0.48\n",
       "\n",
       "[242 rows x 6 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 表8、表9 样品全基因组及各染色体的平均甲基化水平\n",
    "\n",
    "def calc_methylation_level_by_chromosome(sample_name):\n",
    "    input_file=f\"{root_dir}{sample_name}/output/{sample_name}_methylation_coverage_report.txt\"\n",
    "    output_file=f\"{root_dir}report/{sample_name}/Methylation Level Groupp By Chromosome.tsv\"\n",
    "\n",
    "    # 读取表格文件\n",
    "    df = pd.read_csv(input_file, sep=\"\\t\")\n",
    "\n",
    "    # 过滤只保留Chromosome列中以NC开头的行\n",
    "    df = df[df['Chromosome'].str.startswith('NC')]\n",
    "\n",
    "    # 计算每个 context 的甲基化水平\n",
    "    df['MethylationLevel'] = round(df['totalReadsM'] / df['totalReadsN'] * 100, 2)\n",
    "\n",
    "    # 创建以染色体为index、context为表头的透视表\n",
    "    pivot_table = df.pivot(index='Chromosome', columns='Context', values='MethylationLevel')\n",
    "\n",
    "    # 计算每个染色体的总体甲基化水平\n",
    "    overall_methylation = df.groupby('Chromosome').apply(\n",
    "        lambda x: round(x['totalReadsM'].sum() / x['totalReadsN'].sum() * 100, 2),\n",
    "        include_groups=False\n",
    "    ).reset_index()\n",
    "    overall_methylation.columns = ['Chromosome', 'C (%)']\n",
    "\n",
    "    # 将总体甲基化水平加入透视表\n",
    "    pivot_table = overall_methylation.merge(pivot_table, on='Chromosome')\n",
    "\n",
    "    pivot_table=pivot_table.fillna(0)\n",
    "\n",
    "    pivot_table.columns=[\"Chr\",\"C (%)\", \"CG (%)\", \"CHG (%)\", \"CHH (%)\"]\n",
    "\n",
    "    # 保存结果到文件\n",
    "    pivot_table.to_csv(output_file,sep=\"\\t\",index=False)\n",
    "\n",
    "    return pivot_table\n",
    "\n",
    "df_methylation_level_list=[]\n",
    "for sample_name in sample_names:\n",
    "    df_methylation_level=calc_methylation_level_by_chromosome(sample_name)\n",
    "    # df_methylation_level[\"sample_name\"]=sample_name\n",
    "    df_methylation_level.insert(0,\"sample_name\",sample_name)\n",
    "    df_methylation_level_list.append(df_methylation_level)\n",
    "df_methylation_level_list=pd.concat(df_methylation_level_list,axis=0).reset_index(drop=True)\n",
    "df_methylation_level_list.to_csv(f\"{root_dir}report/Methylation Level Groupp By Chromosome.tsv\",sep=\"\\t\",index=False)\n",
    "df_methylation_level_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1abb105d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 表10 表11 样品在全基因组各类型调控元件范围内的甲基化水平\n",
    "# 暂时不做"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "674cc70c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "from matplotlib.ticker import ScalarFormatter,FuncFormatter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ed1b50de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 绘制M-bias图\n",
    "def plot_mbias(mbias_file,output_dir):\n",
    "    # 读取整个文件内容\n",
    "    with open(mbias_file, 'r') as file:\n",
    "        content = file.read()\n",
    "\n",
    "    # 定义正则表达式查找不同上下文表格\n",
    "    # 正则表达式将匹配上下文标题、分隔符、表头和表数据\n",
    "    pattern = r'(\\w+) context \\((\\w+)\\)\\n=+\\n([\\s\\S]+?)(?=\\n[A-Z]|$)'\n",
    "    matches = re.findall(pattern, content)\n",
    "\n",
    "    df = []\n",
    "    # 解析匹配到的表格\n",
    "    for context, read, table_data in matches:\n",
    "        # 将表格转换为 DataFrame\n",
    "        item = pd.read_csv(StringIO(table_data), sep='\\t')\n",
    "        # 添加上下文和read标签列到 DataFrame\n",
    "        item.insert(0,'context',context)\n",
    "        item.insert(1,'read',read)\n",
    "        df.append(item)\n",
    "\n",
    "    df=pd.concat(df)\n",
    "    for read_name in df['read'].unique():\n",
    "        # 创建一个图表\n",
    "        fig, ax1 = plt.subplots(figsize=(8, 4))\n",
    "\n",
    "        # 创建组合字段\n",
    "        df['read_context'] = df['read'] + ' - ' + df['context']\n",
    "\n",
    "        # 绘制甲基化率（左侧 Y 轴）\n",
    "        sns.lineplot(data=df[df[\"read\"]==read_name], x='position', y='% methylation', hue='context', linewidth=3, ax=ax1)\n",
    "        \n",
    "        ax1.grid(True, linestyle='--', linewidth=0.7, color='#dddddd')\n",
    "        ax1.set_xlabel(\"Position in Read [bp]\")\n",
    "        # 设置 x 和 y 轴从 0 开始\n",
    "        ax1.set_xlim(left=0,right=df[\"position\"].max())\n",
    "        ax1.set_ylim(bottom=0,top=100)\n",
    "        # ax1.set_ylabel(\"% Methylation\", color='tab:blue')\n",
    "        # ax1.tick_params(axis='y', labelcolor='tab:blue')\n",
    "        ax1.set_ylabel(\"% Methylation\")\n",
    "\n",
    "        # 创建第二个 Y 轴（右侧）\n",
    "        ax2 = ax1.twinx()  # 共享 x 轴\n",
    "\n",
    "        # 绘制甲基化数量（右侧 Y 轴）\n",
    "        sns.lineplot(data=df[df[\"read\"]==read_name], x='position', y='count methylated', hue='context', linewidth=1, linestyle='--', ax=ax2)\n",
    "\n",
    "        # 设置第二个 Y 轴\n",
    "        # ax2.set_ylabel(\"# Methylation Calls\", color='tab:orange')\n",
    "        # ax2.tick_params(axis='y', labelcolor='tab:orange')\n",
    "        ax2.set_ylabel(\"# Methylation Calls\")\n",
    "        ax2.set_ylim(bottom=0)\n",
    "\n",
    "        # 自定义格式化函数\n",
    "        def format_func(value, tick_number):\n",
    "            if value >= 10_000_000:\n",
    "                return f'{value / 1_000_000:.0f}M'\n",
    "            elif value >= 1_000_000:\n",
    "                return f'{value / 1_000_000:.1f}M'\n",
    "            elif value >= 10_000:\n",
    "                return f'{value / 1_000:.0f}K'\n",
    "            elif value >= 1_000:\n",
    "                return f'{value / 1_000:.1f}K'\n",
    "            else:\n",
    "                return f'{value:.0f}'\n",
    "        ax2.yaxis.set_major_formatter(FuncFormatter(format_func))\n",
    "\n",
    "        read_fulname={\"R1\":\"Read 1\",\"R2\": \"Read 2\"}[read_name]\n",
    "        plt.title(f\"{sample_name} {read_fulname}\")\n",
    "\n",
    "        # 显示图例\n",
    "        lines, labels = ax1.get_legend_handles_labels()\n",
    "        lines2, labels2 = ax2.get_legend_handles_labels()\n",
    "        labels = [label + \" methylation\" for label in labels]\n",
    "        labels2 = [label + \" total calls\" for label in labels2]\n",
    "        ax1.legend(lines + lines2, labels + labels2,loc=\"upper right\")\n",
    "        ax2.get_legend().remove()\n",
    "\n",
    "        plt.savefig(f\"{output_dir}M-bias {read_name}.jpg\",dpi=1000)\n",
    "        # plt.show()\n",
    "        plt.close()\n",
    "for sample_name in sample_names:\n",
    "    mbias_file = f\"{root_dir}{sample_name}/output/bismark_methylation/{sample_name}_1_bismark_bt2_pe.deduplicated.M-bias.txt\"\n",
    "    output_dir=f\"{root_dir}report/{sample_name}/\"\n",
    "    plot_mbias(mbias_file, output_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a74a9c7d",
   "metadata": {},
   "source": [
    "## 3. 甲基化 C碱基中 CG, CHG 与CHH的分布比例\n",
    "mCG，mCHG和mCHH三种碱基类型的构成比例在不同物种中，甚至在同一物种不同样品中都存在很大差异。因此，不同时间、空间、生理条件下的样品会表现出不同的甲基化图谱，各类型mC( mCG、mCHG和mCHH )的数目，及其在全部mC的位点中所占的比例，在一定程度上反映了特定物种的全基因组甲基化图谱的特征。mCG、mCHG和mCHH分别表示表示甲基化CG、甲基化CHG和甲基化CHH。三种碱基类型占比总和为100%，甲基化C鉴定方法依据Lister的文章描述进行。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5b776173",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mCG</th>\n",
       "      <th>mCHG</th>\n",
       "      <th>mCHH</th>\n",
       "      <th>mCG proportion (%)</th>\n",
       "      <th>mCHG proportion (%)</th>\n",
       "      <th>mCHH proportion (%)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample Id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13A</th>\n",
       "      <td>353582218</td>\n",
       "      <td>13466905</td>\n",
       "      <td>43179108</td>\n",
       "      <td>86.19</td>\n",
       "      <td>3.28</td>\n",
       "      <td>10.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32A</th>\n",
       "      <td>380305619</td>\n",
       "      <td>14970374</td>\n",
       "      <td>48448057</td>\n",
       "      <td>85.71</td>\n",
       "      <td>3.37</td>\n",
       "      <td>10.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50A</th>\n",
       "      <td>385841393</td>\n",
       "      <td>15132609</td>\n",
       "      <td>51331234</td>\n",
       "      <td>85.31</td>\n",
       "      <td>3.35</td>\n",
       "      <td>11.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26A</th>\n",
       "      <td>409455527</td>\n",
       "      <td>15141590</td>\n",
       "      <td>49161851</td>\n",
       "      <td>86.43</td>\n",
       "      <td>3.20</td>\n",
       "      <td>10.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28A</th>\n",
       "      <td>369382134</td>\n",
       "      <td>13577988</td>\n",
       "      <td>43129274</td>\n",
       "      <td>86.69</td>\n",
       "      <td>3.19</td>\n",
       "      <td>10.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38A</th>\n",
       "      <td>395482471</td>\n",
       "      <td>15185144</td>\n",
       "      <td>51921364</td>\n",
       "      <td>85.49</td>\n",
       "      <td>3.28</td>\n",
       "      <td>11.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74A</th>\n",
       "      <td>386013982</td>\n",
       "      <td>14606916</td>\n",
       "      <td>49230271</td>\n",
       "      <td>85.81</td>\n",
       "      <td>3.25</td>\n",
       "      <td>10.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25A</th>\n",
       "      <td>463998830</td>\n",
       "      <td>17218295</td>\n",
       "      <td>55282291</td>\n",
       "      <td>86.49</td>\n",
       "      <td>3.21</td>\n",
       "      <td>10.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75A</th>\n",
       "      <td>330761705</td>\n",
       "      <td>12936961</td>\n",
       "      <td>42390800</td>\n",
       "      <td>85.67</td>\n",
       "      <td>3.35</td>\n",
       "      <td>10.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88A</th>\n",
       "      <td>429802771</td>\n",
       "      <td>16072004</td>\n",
       "      <td>51571568</td>\n",
       "      <td>86.40</td>\n",
       "      <td>3.23</td>\n",
       "      <td>10.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97A</th>\n",
       "      <td>412720712</td>\n",
       "      <td>16093115</td>\n",
       "      <td>51514004</td>\n",
       "      <td>85.92</td>\n",
       "      <td>3.35</td>\n",
       "      <td>10.72</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 mCG      mCHG      mCHH  mCG proportion (%)  \\\n",
       "Sample Id                                                      \n",
       "13A        353582218  13466905  43179108               86.19   \n",
       "32A        380305619  14970374  48448057               85.71   \n",
       "50A        385841393  15132609  51331234               85.31   \n",
       "26A        409455527  15141590  49161851               86.43   \n",
       "28A        369382134  13577988  43129274               86.69   \n",
       "38A        395482471  15185144  51921364               85.49   \n",
       "74A        386013982  14606916  49230271               85.81   \n",
       "25A        463998830  17218295  55282291               86.49   \n",
       "75A        330761705  12936961  42390800               85.67   \n",
       "88A        429802771  16072004  51571568               86.40   \n",
       "97A        412720712  16093115  51514004               85.92   \n",
       "\n",
       "           mCHG proportion (%)  mCHH proportion (%)  \n",
       "Sample Id                                            \n",
       "13A                       3.28                10.53  \n",
       "32A                       3.37                10.92  \n",
       "50A                       3.35                11.35  \n",
       "26A                       3.20                10.38  \n",
       "28A                       3.19                10.12  \n",
       "38A                       3.28                11.22  \n",
       "74A                       3.25                10.94  \n",
       "25A                       3.21                10.30  \n",
       "75A                       3.35                10.98  \n",
       "88A                       3.23                10.37  \n",
       "97A                       3.35                10.72  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 表12 mCG、mCHG和mCHH三种类型甲基化胞嘧啶的比例\n",
    "\n",
    "report_table12=[]\n",
    "for sample_name in sample_names:\n",
    "    data={}\n",
    "    with open(f\"{root_dir}{sample_name}/output/bismark_methylation/{sample_name}_1_bismark_bt2_pe.deduplicated_splitting_report.txt\", \"r\") as file:\n",
    "        for line in file:\n",
    "            if line.count(\":\") == 1:   # 匹配包含1个冒号的行\n",
    "                parts = line.split(\":\")\n",
    "                try:\n",
    "                    data[parts[0].strip()]=int(parts[1].strip())\n",
    "                except:\n",
    "                    pass\n",
    "    item={}\n",
    "    item[\"Sample Id\"]=sample_name\n",
    "    item[\"mCG\"]=data[\"Total methylated C's in CpG context\"]\n",
    "    item[\"mCHG\"]=data[\"Total methylated C's in CHG context\"]\n",
    "    item[\"mCHH\"]=data[\"Total methylated C's in CHH context\"]\n",
    "    total_mC=item[\"mCG\"]+item[\"mCHG\"]+item[\"mCHH\"]\n",
    "    item[\"mCG proportion (%)\"]=round(item[\"mCG\"]/total_mC*100,2)\n",
    "    item[\"mCHG proportion (%)\"]=round(item[\"mCHG\"]/total_mC*100,2)\n",
    "    item[\"mCHH proportion (%)\"]=round(item[\"mCHH\"]/total_mC*100,2)\n",
    "\n",
    "    \n",
    "    report_table12.append(item)\n",
    "report_table12=pd.DataFrame(report_table12).set_index(\"Sample Id\")\n",
    "report_table12.to_csv(f\"{root_dir}report/Proportions of Three Types of Methylated Cytosine.tsv\",sep=\"\\t\")\n",
    "\n",
    "report_table12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "113e40db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 图5 不同序列类型甲基化C碱基的分布比例\n",
    "for sample_name,item in report_table12[[\"mCG\",\"mCHG\",\"mCHH\"]].iterrows():\n",
    "    item.plot(kind='pie', autopct='%1.1f%%', figsize=(5,5))\n",
    "    plt.title(f'The proportion of every type mC (Sample: {sample_name})')\n",
    "    plt.legend()\n",
    "    plt.ylabel(\"\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f\"{root_dir}report/{sample_name}/The proportion of every type mC (Sample: {sample_name}).jpg\",dpi=1000)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "286ec487",
   "metadata": {},
   "source": [
    "## 4. 甲基化 CG、CHG和CHH的甲基化水平分布\n",
    "不同类型的C碱基(mCG、mCHG和mCHH )，其甲基化水平在不同物种间，甚至同一物种不同细胞类型不同条件下其甲基化水平都存在差异。此图统计每种类型( CG、CHG和CHH )甲基化C的甲基化水平分布，反映了该物种DNA甲基化特征"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "868c8a36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 绘制甲基化水平分布图（按每10%分组）\n",
    "def bin_methylation_levels(df, range_size=10):\n",
    "    \"\"\"将数据分组到区间\"\"\"\n",
    "    # 创建 bin 的区间\n",
    "    bin_edges = np.arange(0, 101, range_size)\n",
    "    bin_labels = [f\"{i+range_size}\" for i in bin_edges[:-1]]\n",
    "    df['methylation_level_range'] = pd.cut(df['methylation_level'], bins=bin_edges, labels=bin_labels, right=False)\n",
    "\n",
    "    # 计算每个 methylation_level_range 中的 count 总和\n",
    "    range_df = df.groupby(['context', 'methylation_level_range'],observed=False)['count'].sum().reset_index()\n",
    "    return range_df\n",
    "\n",
    "def plot_methylation_level_distribution(sample_name):\n",
    "    input_file=f\"{root_dir}{sample_name}/output/{sample_name}_methylation_distribution_report.txt\"\n",
    "    # 读取TSV格式的输出文件\n",
    "    df = pd.read_csv(input_file, sep='\\t')\n",
    "    df = df[df[\"methylation_level\"]>0]\n",
    "\n",
    "    # 确保数据按 context 和 methylation_level 排序\n",
    "    df.sort_values(by=['context', 'methylation_level'], inplace=True)\n",
    "\n",
    "    # 将数据分组到指定的区间\n",
    "    df = bin_methylation_levels(df, range_size=10)\n",
    "\n",
    "    # 从长格式转为宽格式\n",
    "    df = df.pivot(index='methylation_level_range', columns='context', values='count')\n",
    "    df.columns.name=None\n",
    "    df.index.name=None\n",
    "\n",
    "    # 计算总甲基化数量\n",
    "    df[\"C\"]=df[[\"CG\",\"CHG\",\"CHH\"]].sum(axis=1)\n",
    "    df=df[[\"C\",\"CG\",\"CHG\",\"CHH\"]]\n",
    "    df = df/df.sum()*100  # 计算不同甲基化水平的占比\n",
    "    \n",
    "    # 创建绘图\n",
    "    df.plot(marker='o',figsize=(6, 4))\n",
    "    plt.xlabel('Methylation Level')\n",
    "    plt.ylabel('Percentage (%)')\n",
    "    plt.title(f'Methylation Level Distribution (Sample: {sample_name})')\n",
    "    plt.legend(title='Context')\n",
    "    plt.grid(True, linewidth=1, color='#f6f6f6')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f\"{root_dir}report/{sample_name}/Methylation Level Distribution (Sample: {sample_name}).jpg\",dpi=1000)\n",
    "    plt.close()\n",
    "\n",
    "for sample_name in sample_names:\n",
    "    plot_methylation_level_distribution(sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "d3b4b3d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 绘制甲基化水平累积分布图（按每1%分组）\n",
    "def plot_cumulative_methylation_level_distribution(sample_name):\n",
    "    input_file=f\"{root_dir}{sample_name}/output/{sample_name}_methylation_distribution_report.txt\"\n",
    "    # 读取TSV格式的输出文件\n",
    "    df = pd.read_csv(input_file, sep='\\t')\n",
    "    df = df[df[\"methylation_level\"]>0]\n",
    "\n",
    "    # 确保数据按 context 和 methylation_level 排序\n",
    "    df.sort_values(by=['context', 'methylation_level'], inplace=True)\n",
    "\n",
    "    # 从长格式转为宽格式\n",
    "    df = df.pivot(index='methylation_level', columns='context', values='count')\n",
    "    df.columns.name=None\n",
    "    df.index.name=None\n",
    "\n",
    "    # 计算总甲基化数量\n",
    "    df[\"C\"]=df[[\"CG\",\"CHG\",\"CHH\"]].sum(axis=1)\n",
    "    df=df[[\"C\",\"CG\",\"CHG\",\"CHH\"]]\n",
    "    df = df/df.sum()*100  # 计算不同甲基化水平的占比\n",
    "    df = df.cumsum() # 计算累积分布\n",
    "    \n",
    "    # 创建绘图\n",
    "    df.plot(figsize=(6, 4))\n",
    "    plt.xlabel('Methylation Level')\n",
    "    plt.ylabel('Percentage (%)')\n",
    "    plt.title(f'Cumulative Methylation Level Distribution (Sample: {sample_name})')\n",
    "    plt.legend(title='Context')\n",
    "    plt.grid(True, linewidth=1, color='#f6f6f6')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f\"{root_dir}report/{sample_name}/Cumulative Methylation Level Distribution (Sample: {sample_name}).jpg\",dpi=1000)\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "for sample_name in sample_names:\n",
    "    plot_cumulative_methylation_level_distribution(sample_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6be8c5e8",
   "metadata": {},
   "source": [
    "## 5. 甲基化的CG，CHG，CHH附近碱基的序列特征分析\n",
    "在一些真核生物中，甲基化位点附近碱基的序列特征，对反映甲基化发生的序列偏向有指导意义。为了研究序列特征与甲基化偏向性之间的联系，我们计算了甲基化位点上下游9个碱基 （mC位于第四个碱基）的甲基化百分比。"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c2674a",
   "metadata": {},
   "source": [
    "## 6. 染色体水平的甲基化C碱基密度分布\n",
    "有研究证明非CG型的甲基化与CG型甲基化C的密度有很大的差异。染色体亚端粒区域DNA甲基化水平通常较高，这一现象与端粒长度以及重组有十分重要的作用，此外还有基因表达和蛋白与DNA的相互作用都有紧密联系。 如下图的显示，整体的甲基化水平图谱显示DNA甲基化的密度在整条染色体上变化很大。"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afdced8e",
   "metadata": {},
   "source": [
    "## 7. 基因组的不同区域的甲基化分布特征"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69144d4f",
   "metadata": {},
   "source": [
    "## 8. 基因组不同转录元件中的DNA平均甲基化水平\n",
    "（暂时不绘制）"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a534c00d",
   "metadata": {},
   "source": [
    "## 9. DMR的检测\n",
    "差异甲基化区域（DMRs）是指不同样品中基因组表现出不同的甲基化模式的某些DNA片段。 DMR与遗传印记相关，在个体中表现为与父本或母本的甲基化状态一致。甲基化的等位基因经常表现为沉默状态。亲本与子代甲基化模式的差异常常导致表观遗传缺陷，而人工繁殖技术可能会导致异常甲基化的比例升高，并导致疾病的发生。\n",
    "\n",
    "我们用滑动窗口的方法检测DMR。在两个样品基因组相同位置上寻找包含至少5个CG（或CHG，或CHH）的窗口，比较该窗口在两个样品数据中甲基化水平的差异，寻找在两个样品中甲基化水平有差异的区域。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e582f308",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DMR数据统计（R代码中也实现了这个功能）\n",
    "# def getDMRsStatistics(file_input,file_output):\n",
    "#     columns=[\n",
    "#         \"seqnames\", \"start\", \"end\", \"width\", \"strand\", \"sumReadsM1\", \"sumReadsN1\",\n",
    "#         \"proportion1\", \"sumReadsM2\", \"sumReadsN2\", \"proportion2\", \"cytosinesCount\",\n",
    "#         \"context\", \"direction\", \"pValue\", \"regionType\"\n",
    "#     ]\n",
    "#     df_DMRs=pd.read_csv(file_input,sep=\"\\t\",header=None,names=columns)\n",
    "#     df_DMRs_Statistics=df_DMRs.groupby([\"context\",\"seqnames\"]).size().reset_index(name='DMR number')\n",
    "#     df_DMRs_Statistics.to_csv(file_output,sep=\"\\t\",index=False)\n",
    "#     return df_DMRs_Statistics\n",
    "\n",
    "# file_input = f\"{root_dir}output/DMRsReplicatesBins.chr.txt\"\n",
    "# file_output = f\"{root_dir}report/DMRsummary.tsv\"\n",
    "# df_DMRs_grouped = getDMRsStatistics(file_input,file_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f697c9b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 读取DMR分析结果，并输出对应的基因列表（R代码中也实现了这个功能）\n",
    "# def extract_gene_info(file_path):\n",
    "#     import re\n",
    "\n",
    "#     # 初始化一个空的列表来存储提取的信息\n",
    "#     dmr_genes = []\n",
    "    \n",
    "#     # 打开文件并逐行读取\n",
    "#     with open(file_path, 'r') as file:\n",
    "#         for line in file:\n",
    "#             # 使用正则表达式分别匹配gene_id, gene_type, gene_name\n",
    "#             gene_id_match = re.search(r'gene_id \"([^\"]+)\";', line)\n",
    "#             gene_type_match = re.search(r'gene_type \"([^\"]+)\";', line)\n",
    "#             gene_name_match = re.search(r'gene_name \"([^\"]+)\";', line)\n",
    "            \n",
    "#             # 创建一个字典来存储当前行的匹配结果\n",
    "#             gene_info = {}\n",
    "#             if gene_type_match:\n",
    "#                 gene_info['gene_type'] = gene_type_match.group(1)\n",
    "#             if gene_name_match:\n",
    "#                 gene_info['gene_name'] = gene_name_match.group(1)\n",
    "#             if gene_id_match:\n",
    "#                 gene_info['gene_id'] = gene_id_match.group(1)\n",
    "            \n",
    "#             # 如果三者都匹配到，才加入列表\n",
    "#             if len(gene_info) == 3:\n",
    "#                 dmr_genes.append(gene_info)\n",
    "\n",
    "\n",
    "#     # 将提取的信息转换为DataFrame，并去除重复项，按照gene_type, gene_name, gene_id排序\n",
    "#     dmr_genes = pd.DataFrame(dmr_genes).drop_duplicates().sort_values(by=['gene_type', 'gene_name', 'gene_id']).reset_index(drop=True)\n",
    "\n",
    "#     # 使用 str.split 分割 gene_id 列\n",
    "#     dmr_genes[['gene_id', 'version']] = dmr_genes['gene_id'].str.split('.', 1, expand=True)\n",
    "    \n",
    "#     return dmr_genes\n",
    "\n",
    "# file_path = root_dir+'output/DMR_gene_association.bed'\n",
    "# dmr_genes = extract_gene_info(file_path)\n",
    "# dmr_genes.to_csv(root_dir+'report/DMR_genes.tsv',sep=\"\\t\",index=False)\n",
    "# dmr_genes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python37",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
